### PR TITLE
Revert "Removing unused fetch sub phase processor initialization during fetch… (#12503)"

### PR DESCRIPTION
**Pull request by reta**
This reverts commit da5b2053864f24dcb3b8c555a9fd19e0d91a14e6.

### Description
Reverts https://github.com/opensearch-project/OpenSearch/pull/12503 that introduced regression in 2.13.0. Thanks a lot  @martijnbolhuis for spotting it and supplying the tests, @dblock for working with the contributor.

### Related Issues
Closes https://github.com/opensearch-project/OpenSearch/issues/13467

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [X] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
@jainankitk fyi, @msfroh @dblock please take a look folks (just a revert)

**Review by kotwanikunal — APPROVED** [review]

**Comment by jainankitk** [comment]
@reta - Thank you for this PR. Can you also include the repro as test case for identifying this going forward?

File: server/src/main/java/org/opensearch/search/fetch/subphase/InnerHitsPhase.java [hunk 1]
```diff
@@ -64,7 +64,7 @@ public InnerHitsPhase(FetchPhase fetchPhase) {
 
     @Override
     public FetchSubPhaseProcessor getProcessor(FetchContext searchContext) {
-        if (searchContext.hasInnerHits() == false) {
+        if (searchContext.innerHits() == null) {
             return null;
         }
```
Comment (jainankitk) [inline server/src/main/java/org/opensearch/search/fetch/subphase/InnerHitsPhase.java:69]: We can remove this check as the `return null` is never hit! Probably add some documentation to the effect that we need innerHitsProcessor always included

**Comment by jainankitk** [comment]
> @jainankitk Do you see a fix that avoids the revert and fixes the issue?

We can include the changes from https://github.com/opensearch-project/OpenSearch/pull/13488 as part of this, or drop this and merge the other PR

**Comment by reta** [comment]
> Also, do we need to revert ScriptFieldsPhase change as well? I can also raise PR to just revert inner hits change if you want

I am going to revert the change in full and ask for more tests to be added after

**Comment by jainankitk** [comment]
> I am going to revert the change in full and ask for more tests to be added after

@reta - Can you check #13488 once?

**Comment by dblock** [comment]
> > I am going to revert the change in full and ask for more tests to be added after
> 
> @reta - Can you check #13488 once?

It sounds like @reta and @msfroh would prefer to merge this PR vs. https://github.com/opensearch-project/OpenSearch/pull/13488, I'll let either of them to make that decision and hit the approve/merge buttons. I don't have a strong opinion other than I'm glad we're adding a test for the regression.

**Review by msfroh — APPROVED** [review]

**Pull request merged**